woocommerce/woocommerce
# Tweak useProductTemplate's logic and add new filter to allow changing the product template

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Tweak useProductTemplate's logic and add a new filter to allow changing the product template.

This demand came while migrating an extension where I needed to control the product template loaded depending on the parent's data to ensure backward compatibility with products created in the legacy editor (since new ones have the product template ID persisted to them). 

I've also added a variable product template, which doesn't show in the list. I found that it simplifies the logic and makes it more robust since the special handling starts to get tough when adding new templates supporting variable products. 
Please let me know if any issues with this idea.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate the following plugin: 
[test-change-template.zip](https://github.com/woocommerce/woocommerce/files/15338796/test-change-template.zip)
2. Enable the new product editor at /wp-admin/admin.php?page=wc-settings&tab=advanced&section=features
3. Go to Products > Add   (alternatively, open an already created variable product and skip to step 6)
4. Go to the Variations tab
5. Add variation options
6. Click 'Edit' on a Variation
7. Check the console, the extension installed should log the filter's parameters and change the variation's product template to 'grouped-product-template' (you should see 'This is a grouped product.' in the General tab). 
8. There will be some blocks with errors since some blocks don't play well with variations, but it's OK, since the extension is just to prove the concept (please disable it after testing).
9. Disable the plugin and smoke test creating a simple product, a variable product, and a grouped product.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
